src/rustup/settings.rs

```diff
@@ -0,0 +1,146 @@
+use errors::*;
+use notifications::*;
+use toml_utils::*;
+use utils;
+use toml;
+use std::collections::BTreeMap;
+use std::path::{Path, PathBuf};
+
+pub const SUPPORTED_METADATA_VERSIONS: [&'static str; 2] = ["2", "12"];
+pub const DEFAULT_METADATA_VERSION: &'static str = "12";
+
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SettingsFile(pub PathBuf);
+
+impl SettingsFile {
+    fn write_settings(&self, settings: Settings) -> Result<()> {
+        try!(utils::write_file("settings", &self.0, &settings.stringify()));
+        Ok(())
+    }
+    fn read_settings(&self) -> Result<Settings> {
+        if !utils::is_file(&self.0) {
+            try!(self.write_settings(Default::default()));
+        }
+        let content = try!(utils::read_file("settings", &self.0));
+        Settings::parse(&content)
+    }
+    pub fn with<T, F: FnOnce(&Settings) -> Result<T>>(&self, f: F) -> Result<T> {
+        f(&try!(self.read_settings()))

```

Comment:
I imagine this file is going to get read several times during each run. Can it instead load the file at startup and reuse it?